### PR TITLE
Add new WRF forcing reader in LIS and elev in LDT

### PR DESCRIPTION
**Pull request by karsenau**
A new WRF-based forcing reader has been implemented into
 LIS that supports higher-res, CONUS-wide WRF datasets.

 Also along with the new forcing reader, WRF-based elevation
 terrain map support is also incorporated in LDT and can be
 used with the new WRF reader in LIS.

 New options are highlighted in the "configs" documents.

This applies to feature issue #811 and completes that ticket.

### Testcase

Test cases for this reader are found here:
/discover/nobackup/karsenau/LISF_PR/WRF_ForcReader

In the "LIS_Test" subdirectory, all files are provided, except the parameters and forcing are provided via symbolic line ("input") which points to the top-level LIS data disk.  All paths should be final and the person evaluating this pull-request and test-case, should be able to copy the directory "as-is" to their test directory. Also, sample "target" output and files are provided in the same directory of the test-case run with the new forcing reader (found in the local dir: NoahMP401_WRFforc_NoLR_target).

Finally, the WRF-forcing terrain file also has a reader to support it on the LDT side and can be switched on in LIS by selecting "lapse-rate" in your lis.config file. The LDT testcase for this set up is found here:

/discover/nobackup/karsenau/LISF_PR/WRF_ForcReader/LDT_Test

For any further information or have any questions, please contact @karsenau.

### CODE REVIEW

File: lis/metforcing/WRFoutv2/0Intro_WRFoutv2.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+!-----------------------BEGIN NOTICE -- DO NOT EDIT-----------------------
+! NASA Goddard Space Flight Center Land Information System (LIS) v7.2
+!
+! Copyright (c) 2015 United States Government as represented by the
+! Administrator of the National Aeronautics and Space Administration.
+! All Rights Reserved.
+!-------------------------END NOTICE -- DO NOT EDIT-----------------------
```
Comment (bmcandr) [inline lis/metforcing/WRFoutv2/0Intro_WRFoutv2.txt:7]: Update copyright to 2020 version:

```
!-----------------------BEGIN NOTICE -- DO NOT EDIT-----------------------
! NASA Goddard Space Flight Center
! Land Information System Framework (LISF)
! Version 7.3
!
! Copyright (c) 2020 United States Government as represented by the
! Administrator of the National Aeronautics and Space Administration.
! All Rights Reserved.
!-------------------------END NOTICE -- DO NOT EDIT-----------------------
```

File: lis/metforcing/WRFoutv2/WRFoutv2_forcingMod.F90 [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+    enddo
+
+    !tmp     <--> LIS_forc%metdata1(1,:)
+    !q2      <--> LIS_forc%metdata1(2,:)
+    !swd     <--> LIS_forc%metdata1(3,:)
+    !lwd     <--> LIS_forc%metdata1(4,:)
+    !uwind   <--> LIS_forc%metdata1(5,:)
+    !vwind   <--> LIS_forc%metdata1(6,:)
+    !psurf   <--> LIS_forc%metdata1(7,:)
+    !pcp     <--> LIS_forc%metdata1(8,:)
```
Comment (bmcandr) [inline lis/metforcing/WRFoutv2/WRFoutv2_forcingMod.F90:365]: Remove if not needed.

File: lis/metforcing/WRFoutv2/read_WRFoutv2.F90 [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+         status = nf90_inq_varid(ncid, "XLONG", lonid)
+         status = nf90_get_var(ncid, lonid, lon, &
+                       start=(/1,1/), &
+                       count=(/WRFoutv2_struc(n)%nc,WRFoutv2_struc(n)%nr/))
+       endif
+
+!  Note: datain --> datain(y,x) when read in
+!       print *, trim(WRFoutv2_fv(v))//" ... datain(5,9): ",datain(5,9)   ! t=4 ==> T2=295.045
+!       print *, trim(WRFoutv2_fv(v))//" ... datain(1,2): ",datain(1,2)    ! t=1 ==> T2=295.2789
+!       print *, trim(WRFoutv2_fv(v))//" ... datain(1,2): ",datain(1,2)    ! t=2 ==> T2=295.283
```
Comment (bmcandr) [inline lis/metforcing/WRFoutv2/read_WRFoutv2.F90:255]: Remove if not needed.

File: lis/metforcing/WRFoutv2/read_WRFoutv2.F90 [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+
+     do v=1,NF      ! do not process convective precip field, which is set to 0
+      
+        !== Transferring current data to 1-D array for interpolation
+        c=0
+        do j=1,WRFoutv2_struc(n)%nr
+           do i=1,WRFoutv2_struc(n)%nc
+              c = c + 1
+              f(c) = temp2WRFoutv2(i,j,v)
+!              if(v.eq.1 .and. f(c).ne.-9999.0) write(LIS_logunit,*) c,f(c)
```
Comment (bmcandr) [inline lis/metforcing/WRFoutv2/read_WRFoutv2.F90:335]: Remove if not needed.

File: lis/metforcing/WRFoutv2/read_WRFoutv2.F90 [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+                WRFoutv2_struc(n)%mi,mo,&
+                LIS_domain(n)%lat, LIS_domain(n)%lon,&
+                WRFoutv2_struc(n)%n113,LIS_rc%udef,iret)
+
+        end select
+
+    
+        !== Convert data to original 3D array & a 2D array to 
+        !== Fill in of missing points due to geography difference  
+!        tg = -9999.0
```
Comment (bmcandr) [inline lis/metforcing/WRFoutv2/read_WRFoutv2.F90:382]: Remove if not needed.

File: lis/metforcing/WRFoutv2/readconfig_WRFoutv2.F90 [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  call LIS_verify(rc, 'WRF output v2 forcing directory: not defined')
+  do n=1,LIS_rc%nnest    
+     call ESMF_ConfigGetAttribute(LIS_config,WRFoutv2_struc(n)%WRFoutv2dir,rc=rc)
+  enddo
+
+!  call ESMF_ConfigFindLabel(LIS_config,"WRF nest id:",rc=rc)
+  do n=1,LIS_rc%nnest    
+     WRFoutv2_struc(n)%nest_id = 1
+!     call ESMF_ConfigGetAttribute(LIS_config,WRFoutv2_struc(n)%nest_id,rc=rc)
+  enddo
```
Comment (bmcandr) [inline lis/metforcing/WRFoutv2/readconfig_WRFoutv2.F90:48]: Should `WRF nest id:` be read from the config? It is listed in `lis.config.adoc` as an option.

**Review by bmcandr — CHANGES_REQUESTED** [review]
Hi @karsenau,

The code looks good overall, but I have a few minor cleanup requests.

I'm also seeing some slight differences between my output and the target for the LIS *LRcorr* test. Can you take a peek at the output and see if this looks good to you? See the output in `*LRcorr_test` in `/discover/nobackup/projects/lis/PR_TESTCASES/818/testing/WRF_ForcReader/LIS_Test`.

**Comment by bmcandr** [comment]
Not sure why the first comment shows up twice... I tried out the inline review feature and must have done something wrong!

**Comment by karsenau** [comment]
Hi @bmcandr:

Thank you so much for the thorough review of the code and testcase results.

I just checked your testcase outputs for the "lapse-rate" correction case against the "target" output (that I generated), and I didn't see much difference, checked for date/timestamps: 2004101903, 2004102009

Was there a particular date/time that you saw differences?

Thanks!

**Review by karsenau — COMMENTED** [review]
Updates made, per @bmcandr's review and suggestions.

**Review by bmcandr — CHANGES_REQUESTED** [review]
> Hi @bmcandr:
> 
> Thank you so much for the thorough review of the code and testcase results.
> 
> I just checked your testcase outputs for the "lapse-rate" correction case against the "target" output (that I generated), and I didn't see much difference, checked for date/timestamps: 2004101903, 2004102009
> 
> Was there a particular date/time that you saw differences?
> 
> Thanks!

Hi @karsenau,

I'm seeing some small differences between the target and test outputs for several dates when I compare them with `nccmp -dsf`. They are very small so I just wanted to make sure they are OK with you. If you'd like to see what I'm talking about, `cd` to my testing directory and run:

```shell
for TARGET in $(find NoahMP401_WRFforc_LRcorr_target -name "LIS_HIST*.nc"); do TEST=${TARGET/target/test}; echo $TEST; nccmp -dsf $TARGET $TEST; done
```

----

I have a few more revisions to request on the files added for the Alaska domain. Also, is there a testcase for the AK domain reader?

**Comment by bmcandr** [comment]
After speaking with @karsenau, the requested changes to the Alaska domain reader will be deferred until a later pull request.

Otherwise, the code looks good and I was able to reproduce the results with both the Intel19 and GNU9 compilers. Merging.

**Review by bmcandr — APPROVED** [review]

**Comment by karsenau** [comment]
The new WRF Alaska (AK) reader in LIS and its elevation file reader in LDT are now ready, along with test cases for LDT and LIS (with and without topographic corrections).

The test case files can be found here:

LDT:  /discover/nobackup/karsenau/LISF_PR/WRF_AKReader/LDT_Test
LIS:  /discover/nobackup/karsenau/LISF_PR/WRF_AKReader/LIS_Test

This PR addresses feature issue #812 .

**Pull request merged**